osu.Game/Overlays/Comments/CommentEditor.cs

```diff
@@ -134,19 +162,38 @@ private void load(OverlayColourProvider colourProvider)
             });
 
             TextBox.OnCommit += (_, _) => commitButton.TriggerClick();
+            User.BindTo(API.LocalUser);
         }
 
         protected override void LoadComplete()
         {
             base.LoadComplete();
             Current.BindValueChanged(_ => updateCommitButtonState(), true);
+            User.BindValueChanged(_ => updateStateForLoggedIn(), true);
```

Comment:
Is there a reason we're binding to `User` to get the api's `State`?

```diff
diff --git a/osu.Game/Overlays/Comments/CommentEditor.cs b/osu.Game/Overlays/Comments/CommentEditor.cs
index 8dbafc0ea8..05bdac5966 100644
--- a/osu.Game/Overlays/Comments/CommentEditor.cs
+++ b/osu.Game/Overlays/Comments/CommentEditor.cs
@@ -14,7 +14,6 @@
 using osu.Game.Graphics.UserInterface;
 using osu.Game.Graphics.UserInterfaceV2;
 using osu.Game.Online.API;
-using osu.Game.Online.API.Requests.Responses;
 using osuTK;
 using osuTK.Graphics;
 
@@ -36,14 +35,14 @@ public abstract partial class CommentEditor : CompositeDrawable
 
         protected TextBox TextBox { get; private set; } = null!;
 
-        protected readonly IBindable<APIUser> User = new Bindable<APIUser>();
-
         [Resolved]
         protected IAPIProvider API { get; private set; } = null!;
 
         [Resolved]
         private LoginOverlay? loginOverlay { get; set; }
 
+        private readonly IBindable<APIState> apiState = new Bindable<APIState>();
+
         /// <summary>
         /// Returns the text content of the main action button.
         /// When <paramref name="isLoggedIn"/> is <see langword="true"/>, the text will apply to a button that posts a comment.
@@ -162,14 +161,14 @@ private void load(OverlayColourProvider colourProvider)
             });
 
             TextBox.OnCommit += (_, _) => commitButton.TriggerClick();
-            User.BindTo(API.LocalUser);
+            apiState.BindTo(API.State);
         }
 
         protected override void LoadComplete()
         {
             base.LoadComplete();
             Current.BindValueChanged(_ => updateCommitButtonState(), true);
-            User.BindValueChanged(_ => updateStateForLoggedIn(), true);
+            apiState.BindValueChanged(updateStateForLoggedIn, true);
         }
 
         protected abstract void OnCommit(string text);
@@ -177,12 +176,14 @@ protected override void LoadComplete()
         private void updateCommitButtonState() =>
             commitButton.Enabled.Value = loadingSpinner.State.Value == Visibility.Hidden && !string.IsNullOrEmpty(Current.Value);
 
-        private void updateStateForLoggedIn()
+        private void updateStateForLoggedIn(ValueChangedEvent<APIState> state)
         {
-            TextBox.PlaceholderText = GetPlaceholderText(API.IsLoggedIn);
-            TextBox.ReadOnly = !API.IsLoggedIn;
+            bool isAvailable = state.NewValue > APIState.Offline;
+
+            TextBox.PlaceholderText = GetPlaceholderText(isAvailable);
+            TextBox.ReadOnly = !isAvailable;
 
-            if (API.IsLoggedIn)
+            if (isAvailable)
             {
                 commitButton.Show();
                 logInButton.Hide();

```

Reply:
I didn't put too much thought into it. My thinking was that changes in `User` correspond to changes between just `APIState.Offline` and `APIState.Online`, but binding to `state` may fire for other state changes where the user hasn't actually changed: for instance, maybe your wifi going could result in your api state changing `APIState.Failing` (I don't know whether this is actually how `state` works).

Of course, since the ui only checks whether the state is offline or "not offline" (online), no harm is done by firing the update function additional times for other states. It doesn't do any good, either, since we only care whether the user is the guest user or not.

If binding to `state` is generally considered better practice for checking "is the user logged in", then let's do that.

Reply:
Yep, using state is the correct way. Feels weird to be using an arbitrary piece of information (`User`) to trigger an event then read the api state from a different field, after all.